### PR TITLE
Allow static inheritence for ScriptModules

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#20503 Allow static inheritence for ScriptModules**

Summary: Previously, we registered script modules of parent classes during their init methods, making it impossible
to correctly 'override' a method or leave it blank in a subclass. This diff adds the functionality by waiting
until the child init is finished before registering all the methods.

Test Plan: python test/test_jit.py

Differential Revision: [D15341555](https://our.internmc.facebook.com/intern/diff/D15341555)

### CODE REVIEW

File: test/test_jit.py [hunk 1]
```diff
@@ -3294,6 +3294,29 @@ def foo(x):
...
+            @torch.jit.script_method
+            def bar(self, x):
+                return x * x
+
+        with self.assertRaisesRegex(RuntimeError, 'attribute'):
+            A() # cannot use because bar is not defined
+        v = torch.rand(3, 4)
+        b = B()
+        self.assertEqual(b(v), v + v * v)
+>>>>>>> Allow static inheritence for ScriptModules
```
Comment (eellison) [inline test/test_jit.py]: looks like merge error here

**Review by eellison — APPROVED** [review]
Looks good!

Could you add a test for a module inheritance hierarchy of 3 ? It's not clear to me that we're currently iterating through bases in the right order, since the current test only has a single base.

**Comment by TheCodez** [comment]
Is it possible to call the super class forward method in the derived class forward method?

**Comment by zdevito** [comment]
super is not support and it will take some time before we can support it.

**Pull request closed without merging**